File: app/src/main/java/at/msd/friehs_bicha/cdcsvparser/AssetsFilterActivity.java
package at.msd.friehs_bicha.cdcsvparser;

import static java.lang.Thread.sleep;
import static at.msd.friehs_bicha.cdcsvparser.util.PreferenceHelper.getUseAndroidDB;

import android.content.Context;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import at.msd.friehs_bicha.cdcsvparser.App.AppType;
import at.msd.friehs_bicha.cdcsvparser.general.AppModel;
import at.msd.friehs_bicha.cdcsvparser.transactions.Transaction;
import at.msd.friehs_bicha.cdcsvparser.util.PreferenceHelper;
import at.msd.friehs_bicha.cdcsvparser.wallet.CroCardWallet;
import at.msd.friehs_bicha.cdcsvparser.wallet.Wallet;

public class AssetsFilterActivity extends AppCompatActivity {

    AppModel appModel;
    Context context;

    /**
     * sets the spinner and context the first time
     */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_assets_filter);
        // calling the action bar
        ActionBar actionBar = getSupportActionBar();

        context = getApplicationContext();

        // showing the back button in action bar
        actionBar.setDisplayHomeAsUpEnabled(true);

        Spinner dropdown = findViewById(R.id.asset_spinner);
        //create a list of items for the spinner.

        getAppModel();

        //test Internet connection
        //testConnection();

        while (!appModel.isRunning) {
            try {
                sleep(500);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }

        // make List with all Wallets
        String[] items = getWalletNames();
        //create an adapter to describe how the items are displayed
        ArrayAdapter<String> assetNamesAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_dropdown_item, items);
        //set the spinners adapter to the previously created one.
        dropdown.setAdapter(assetNamesAdapter);

        //get the specific wallet
        Object indexObj = dropdown.getSelectedItem();
        if (indexObj == null) {
            return;
        }
        Wallet specificWallet = appModel.txApp.wallets.get(appModel.txApp.wallets.get(0).getWallet(indexObj.toString()));

        // display prices
        displayInformation(specificWallet, findViewById(R.id.all_regarding_tx));

        displayTxs(specificWallet);

        //if spinner item gets changed
        dropdown.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parentView, View selectedItemView, int position, long id) {
                //get the specific wallet
                Wallet specificWallet = appModel.txApp.wallets.get(appModel.txApp.wallets.get(0).getWallet(dropdown.getSelectedItem().toString()));

                //display Transactions
                displayTxs(specificWallet);
                //display prices
                displayInformation(specificWallet, findViewById(R.id.all_regarding_tx));
            }

            @Override
            public void onNothingSelected(AdapterView<?> parentView) {
            }
        });


    }


    private void getAppModel() {
        if ((PreferenceHelper.getSelectedType(getApplicationContext()) == AppType.CdCsvParser) && getUseAndroidDB(getApplicationContext())) {
            appModel = new AppModel(PreferenceHelper.getSelectedType(this), PreferenceHelper.getUseStrictType(this), getApplicationContext());
        } else {
            appModel = (AppModel) getIntent().getExtras().get("AppModel");
        }
    }


    /**
     * Gets the wallet names for the selected appType
     *
     * @return the wallet names as String[]
     */
    @NonNull
    private String[] getWalletNames() {
        ArrayList<String> wallets = new ArrayList<>();
        switch (appModel.appType) {
            case CdCsvParser:
                appModel.txApp.wallets.forEach(wallet -> wallets.add(wallet.getCurrencyType()));
                wallets.remove("EUR");
                break;
            case CroCard:
                appModel.txApp.wallets.forEach(wallet -> wallets.add(((CroCardWallet) wallet).getTransactionType()));
                break;
            default:
                wallets.add("This should not happen");
        }

        String[] items = wallets.toArray(new String[0]);
        return items;
    }


    /**
     * Checks if there is an internet connection
     */
    private void testConnection() {
        Thread t1 = new Thread(() -> {
            try {
                appModel.getValueOfAssets();
                AppModel.asset.isRunning = true;
            } catch (Exception e) {
                System.out.println("no internet connection");
            }
        });
        t1.start();
    }


    /**
     * Displays the prices of specificWallet
     *
     * @param specificWallet   which should be displayed
     * @param all_regarding_tx the TextView which should be set
     */
    private void displayInformation(Wallet specificWallet, TextView all_regarding_tx) {
        all_regarding_tx.setText("All transactions regarding " + specificWallet.getCurrencyType());
        if (appModel.appType == AppType.CroCard) {
            all_regarding_tx.setText("All transactions regarding " + ((CroCardWallet) specificWallet).getTransactionType());
        }

        //get and set prices
        Thread t = new Thread(() -> displayTexts(appModel.getAssetMap(specificWallet)));
        t.start();

    }


    /**
     * Displays the prices of specificWallet
     *
     * @param texts the Map<String, String> which should be displayed with id of View and text to set pairs
     */
    private void displayTexts(Map<String, String> texts) {
        texts.forEach((key, value) -> {
            TextView textView = findViewById(getResources().getIdentifier(key, "id", getPackageName()));
            if (value == null) {
                AssetsFilterActivity.this.runOnUiThread(() -> textView.setVisibility(View.INVISIBLE));
            } else {
                AssetsFilterActivity.this.runOnUiThread(() -> textView.setText(value));
            }
        });
    }


    /**
     * Displays the Transactions of specificWallet
     *
     * @param specificWallet the CDCWallet which should be displayed
     */
    private void displayTxs(Wallet specificWallet) {
        // Get a reference to the ListView
        ListView listView = findViewById(R.id.lv_txs);

        List<Transaction> transactions = specificWallet.getTransactions();

        // Create an adapter for the ListView
        ArrayAdapter<Transaction> adapterLV = new ArrayAdapter<>(context, android.R.layout.simple_list_item_1, transactions);

        // Set the adapter on the ListView
        listView.setAdapter(adapterLV);
    }


    /**
     * Set the back button in action bar
     */
    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            this.finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/at/msd/friehs_bicha/cdcsvparser/AssetsFilterActivity.kt
package at.msd.friehs_bicha.cdcsvparser

import android.content.Context
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import at.msd.friehs_bicha.cdcsvparser.Core.CoreService
import at.msd.friehs_bicha.cdcsvparser.logging.FileLog
import at.msd.friehs_bicha.cdcsvparser.transactions.Transaction
import at.msd.friehs_bicha.cdcsvparser.ui.fragments.TransactionFragment
import at.msd.friehs_bicha.cdcsvparser.wallet.Wallet

class AssetsFilterActivity : AppCompatActivity() {
    var context: Context? = null
    private var walletList: MutableList<Wallet> = mutableListOf()
    private var walletNamesArray: Array<String?> = arrayOf()

    /**
     * sets the spinner and context the first time
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_assets_filter)
        // calling the action bar
        val actionBar = supportActionBar
        context = applicationContext

        // showing the back button in action bar
        actionBar!!.setDisplayHomeAsUpEnabled(true)
        val dropdown = findViewById<Spinner>(R.id.asset_spinner)
        // make List with all Wallets
        val items = CoreService.walletNames.value ?: arrayOf()
        walletList = CoreService.walletsLiveData.value ?: mutableListOf()

        if (items.isEmpty()) {
            FileLog.e("AssetsFilterActivity", "items is empty")
            return
        }

        walletNamesArray = items.copyOf()
        //create an adapter to describe how the items are displayed
        val assetNamesAdapter =
            ArrayAdapter(this, android.R.layout.simple_spinner_dropdown_item, items)
        //set the spinners adapter to the previously created one.
        dropdown.adapter = assetNamesAdapter

        //get the specific wallet
        val indexObj = dropdown.selectedItem ?: return
        var specificWallet = walletList.find { it.getTypeString() == indexObj }

        val intentWalletId = intent.extras?.getInt("walletID", -1)
        if (intentWalletId != null && intentWalletId != -1) {
            val wallet = walletList.find { it.walletId == intentWalletId }
            if (wallet != null) {
                specificWallet = wallet
            } else {
                FileLog.e("AssetsFilterActivity", "Wallet not found")
            }
            val index = walletList.indexOf(wallet)
            dropdown.setSelection(index)
        }


        // display prices
        displayInformation(specificWallet, findViewById(R.id.all_regarding_tx))

        //displays transactions
        if (specificWallet != null) supportFragmentManager.beginTransaction()
            .replace(
                R.id.fragment_container,
                TransactionFragment(specificWallet.transactions as ArrayList<Transaction>)
            )
            .commit()

        //if spinner item gets changed
        dropdown.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(
                parentView: AdapterView<*>?,
                selectedItemView: View,
                position: Int,
                id: Long
            ) {
                //get the specific wallet
                val specificWallet =
                    walletList[position]

                //display Transactions
                supportFragmentManager.beginTransaction()
                    .replace(
                        R.id.fragment_container,
                        TransactionFragment(specificWallet.transactions as ArrayList<Transaction>)
                    )
                    .commit()
                //display prices
                displayInformation(specificWallet, findViewById(R.id.all_regarding_tx))
            }

            override fun onNothingSelected(parentView: AdapterView<*>?) {}
        }
    }


    /**
     * Displays the prices of specificWallet
     *
     * @param specificWallet   which should be displayed
     * @param all_regarding_tx the TextView which should be set
     */
    private fun displayInformation(specificWallet: Wallet?, all_regarding_tx: TextView) {

        if (specificWallet == null) {
            FileLog.e("AssetsFilterActivity", "specificWallet is null")
            return
        }

        all_regarding_tx.text =
            resources.getString(R.string.all_transactions_regarding) + specificWallet.getTypeString()


        //get and set prices
        val t = Thread { displayTexts(CoreService.getAssetMap(specificWallet.walletId)) }
        t.start()
    }

    /**
     * Displays the prices of specificWallet
     *
     * @param texts the Map<String></String>, String> which should be displayed with id of View and text to set pairs
     */
    private fun displayTexts(texts: Map<String, String?>) {
        texts.forEach { (key: String?, value: String?) ->
            val textView = findViewById<TextView>(resources.getIdentifier(key, "id", packageName))
            if (value == null) {
                runOnUiThread { textView.visibility = View.INVISIBLE }
            } else {
                runOnUiThread { textView.text = value }
            }
        }
    }

    /**
     * Set the back button in action bar
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}